modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsRepository.java

```diff
@@ -528,47 +444,69 @@ public void updateObsolescenceInfo(Map<StatisticsObjectConfiguration, Set<Intege
 
         for (Map.Entry<StatisticsKey, Set<Integer>> objDeleted : deleted.entrySet())
             store.clearObsolescenceInfo(objDeleted.getKey(), objDeleted.getValue());
+
+        fitObsolescenceInfo(cfg);
     }
 
     /**
-     * Load or update obsolescence info cache to fit specified cfg.
+     * Check store to clean unnecessary records.
      *
      * @param cfg Map object statistics configuration to primary partitions set.
      */
-    public synchronized void checkObsolescenceInfo(Map<StatisticsObjectConfiguration, Set<Integer>> cfg) {
-        if (!started.compareAndSet(false, true))
-            loadObsolescenceInfo(cfg);
-        else
-            updateObsolescenceInfo(cfg);
+    private void fitObsolescenceInfo(Map<StatisticsObjectConfiguration, Set<Integer>> cfg) {
```

Comment:
Why don't use BitSet instead of collections of boxed Integers.

Reply:
Yes, it's better to choose an optimal data structure. Removed here.